### DISCUSSION TITLE
How-to testcontainers (with PostgreSQL example)

### DISCUSSION THREAD

**Discussion by langovoi** [adonisjs/core#4453] [Cookbooks]
I searched across Adonis documentation and discussions on Github about using of @testcontainers, but was surprised not find anything.
@testcontainers is a good tool, which allows to have database and not only for tests. For local and same way for CI.

So I figured out how to use it with Adonis and want to share. Guide is focused on PostgreSQL, but sure it can be simply adopted for another testcontainer. And sure it isn't the best way how to use it, so feedback is welcome.

First of all let's create initial point for any testcontainers:
1) create empty file `start/testcontainers.ts`;
2) add `await import('#start/testcontainers')` in `bin/test.ts` before `await import('#start/env')`;

So now we have point where is possible to initialise any testcontainer — `start/testcontainers.ts`.

For example we need PostgreSQL for our tests:
1) `npm i @testcontainers/postgresql --save-dev`;
2) populate `start/testcontainers.ts` with next content:
```typescript
import { PostgreSqlContainer } from '@testcontainers/postgresql'

const postgresContainer = await new PostgreSqlContainer('postgres:15.6-alpine').start()
// here we inject env for `start/env`, which then be used in `config/database.ts`.
process.env.PG_HOST = postgresContainer.getHost()
process.env.PG_PORT = String(postgresContainer.getPort())
process.env.PG_USER = postgresContainer.getUsername()
process.env.PG_DB_NAME = postgresContainer.getDatabase()
process.env.PG_PASSWORD = postgresContainer.getPassword()
```
3) in `tests/bootstrap.ts` in `configureSuite` add next content:
```typescript
  suite.onGroup((group) => {
    group.each.setup(() => {
      return testUtils.db().migrate()
    })
  })
```

Here you can find more testcontainers for Node.js — https://testcontainers.com/modules/?language=nodejs

And if you use Colima, Podman or Rancher take a look to next article: https://node.testcontainers.org/supported-container-runtimes/

**Comment by thetutlage** [comment]
Looks good. I have never used them, but pinging @RomainLanz as he shared test containers with me some time back.

**Reply by RomainLanz** [reply]
Yeah, I personally start them inside the test suite or directly in the test file, depending on the needs.

Example [here](https://github.com/RomainLanz/bus/blob/main/tests/drivers/redis_transport.spec.ts#L18):

```ts
import { test } from '@japa/runner'
import { RedisContainer, type StartedRedisContainer } from '@testcontainers/redis'

test.group('Redis Bus', (group) => {
  let container: StartedRedisContainer

  group.setup(async () => {
    container = await new RedisContainer().start()

    return async () => {
      await container.stop()
    }
  })
})
```

> but was surprised not find anything.

Because there is nothing to be done specific to AdonisJS to make it work.

**Reply by langovoi** [reply]
> Because there is nothing to be done specific to AdonisJS to make it work.

I mean there are no guides with examples and best practices, which could be very helpful to start. So I wrote one and published here.

**Reply by RomainLanz** [reply]
> I mean there are no guides with examples and best practices, which could be very helpful to start. So I wrote one and published here.

And thanks for writing that guide. It is great! 🙌🏻 

I was just explaining why none was in the documentation already. 😄
